src/main/src/main/java/org/geoserver/logging/LoggingUtils.java

```diff
@@ -114,30 +318,222 @@ public static void configureGeoServerLogging(
-                if (curApp instanceof org.apache.log4j.ConsoleAppender) {
-                    org.apache.log4j.Logger.getRootLogger().removeAppender(curApp);
+            for (Appender check : configuration.getAppenders().values()) {
+                if (check instanceof ConsoleAppender) {
+                    configuration.getAppenders().values().remove(check);
+                    reloadRequired = true;
                 }
             }
         }
+        return reloadRequired;
+    }
+
+    /**
+     * Configure via Log4J API.
+     *
+     * <p>While post-processing generating configuration can detect and fix most problems we can
+     * take the time to respect:
+     *
+     * <ul>
+     *   <li>suppressFileLogging: When {@code true}, remove {@code log4j.appender.geoserverlogfile}
+     *   <li>logFilename: added as {@code log4j.appender.geoserverlogfile.File}
+     * </ul>
+     *
+     * This preserves GeoServer 2.20.x Log4J 1.2 API workflow.
+     *
+     * @param configResource log4j properties file
+     * @param noFileLogging True to disable and/or remove all file appenders
+     * @param noConsoleLogging True to disable and/or remove all console appenders
+     * @param logFileName Logfile output location
+     * @return true for successful configuration
+     */
+    static boolean configureFromLog4j2(
+            Resource configResource,
+            final boolean noFileLogging,
+            final boolean noConsoleLogging,
+            final String logFileName) {
+
+        String extension = Paths.extension(configResource.path());
+
+        @SuppressWarnings({
+            "resource",
+            "PMD.CloseResource"
+        }) // current context, no need to enforce AutoClosable
+        LoggerContext loggerContext = (LoggerContext) LogManager.getContext(true);
 
-        // add the appenders we saved above
-        for (Appender appender : appenders) {
-            LogManager.getRootLogger().addAppender(appender);
+        try {
+            URI configLocation = Resources.file(configResource).toURI();
+
+            // Use log4j 1.2 API to reset
+            // org.apache.log4j.LogManager.resetConfiguration();
+
+            // And configure from new location
+            loggerContext.setName(configResource.name());
+
+            if (extension.equalsIgnoreCase("xml")) {
+                ConfigurationSource source = ConfigurationSource.fromUri(configLocation);
+                GeoServerXMLConfiguration configuration =
+                        new GeoServerXMLConfiguration(loggerContext, source) {
+                            @Override
+                            public void setup() {
+                                this.loggingLocation = logFileName;
+                                this.suppressFileLogging = noFileLogging;
+                                this.suppressStdOutLogging = noConsoleLogging;
+                                super.setup();
+                            }
+                        };
```

Comment:
Why are you creating an anonymous `GeoServerXMLConfiguration` child class instead of having a constructor  with `logFileName`, `noFileLogging` and `noConsoleLogging` in `GeoServerXMLConfiguration`?